### PR TITLE
add automatic language content switching

### PR DESCRIPTION
**Pull request by belljohan3**

### CODE REVIEW

**Review by ivanslabbert — CHANGES_REQUESTED** [review]

File: app/src/pages/SingleContent.vue [hunk 1]
```diff
@@ -22,9 +23,30 @@ const content = db.getBySlugAsRef<ContentDto>(props.slug);
...
+        } else {
+            if (appLanguageAsRef.value?._id != content.value.language) {
+                const contentDocs = await db.whereParent(content.value.parentId);
+                const preferred = contentDocs.find(
+                    (c) => c.language == appLanguageAsRef.value?._id,
+                );
+                if (preferred) {
+                    await router.push({ name: "content", params: { slug: preferred.slug } });
+                    content.value = preferred;
+                } else {
```
[the remark this answers is not in the document]
Comment (belljohan3) [reply app/src/pages/SingleContent.vue:39]: The ```else``` here is required. Otherwise it will show the notification even if the content is not found.

[the remark this answers is not in the document]
Comment (laurensgroeneveld) [reply app/src/pages/SingleContent.vue]: What Ivan means is that you can place an empty `return;` statement inside the `if`. Then the `else` is not needed. I also think that is a nicer way of doing it.

File: app/src/pages/SingleContent.vue [hunk 2]
```diff
@@ -22,9 +23,28 @@ const content = db.getBySlugAsRef<ContentDto>(props.slug);
...
 watch(
-    content,
+    [content, appLanguageAsRef],
     async () => {
-        if (!content.value) return;
+        if (!content.value) {
+            return;
+        }
+
+        if (appLanguageAsRef.value?._id != content.value.language) {
```
Comment (laurensgroeneveld) [inline app/src/pages/SingleContent.vue:32]: I don't think this should be in the same watcher as the one for `content`. If I am in an English article that doesn't have a Swahili translation, and change my language to Swahili, I get the notification. But when I reload the page, I get the notification twice, because this watcher triggers everytime either `content` or `appLanguageAsRef` changes. So I think at the very least this should be in it's own watcher.

Comment (laurensgroeneveld) [reply app/src/pages/SingleContent.vue:32]: Although this issue is also something else: the notification will now always show, even if I'm not switching language. That could be the intended behaviour, just want to double check that we do this deliberately.

File: app/src/pages/SingleContent.vue [hunk 3]
```diff
@@ -22,9 +23,28 @@ const content = db.getBySlugAsRef<ContentDto>(props.slug);
...
+            const contentDocs = await db.whereParent(content.value.parentId);
+            const preferred = contentDocs.find((c) => c.language == appLanguageAsRef.value?._id);
+
+            if (preferred) {
+                await router.push({ name: "content", params: { slug: preferred.slug } });
```
Comment (laurensgroeneveld) [inline app/src/pages/SingleContent.vue:37]: Instead of `router.push`, `router.replace` might be a good option as well. It will replace the current URL, so when you click back in the browser, you don't go back to the "wrong" language, but instead to the page before that, which might be the homepage.

File: app/src/pages/SingleContent.vue [hunk 4]
```diff
@@ -22,9 +23,28 @@ const content = db.getBySlugAsRef<ContentDto>(props.slug);
...
+                content.value = preferred;
```
Comment (laurensgroeneveld) [inline app/src/pages/SingleContent.vue:38]: This is needed because routing to the same route will not remount the component but keep using the same instance. For clarity, I would suggest to place this one line up, above the `router` call.

File: app/src/pages/SingleContent.vue [hunk 5]
```diff
@@ -22,9 +23,28 @@ const content = db.getBySlugAsRef<ContentDto>(props.slug);
...
+            } else {
+                useNotificationStore().addNotification({
```
Comment (laurensgroeneveld) [inline app/src/pages/SingleContent.vue:40]: Can we get a test for this?

**Review by laurensgroeneveld — CHANGES_REQUESTED** [review]
Works nicely, just have a few minor comments 👍

**Comment by belljohan3** [comment]
I am struggling to write test for automatic language content switching. @laurensgroeneveld & @ivanslabbert

**Comment by ivanslabbert** [comment]
> I am struggling to write test for automatic language content switching. @laurensgroeneveld & @ivanslabbert

Perhaps write a test where you open an article in French (with the preferred language set to English), and wait to see if the English article title is in wrapper.text()?

File: app/src/pages/SingleContent.spec.ts [hunk 6]
```diff
@@ -126,4 +150,31 @@ describe("SinglePost", () => {
...
+            },
+        });
+
+        // Wait for the initial content to load
+        await waitForExpect(() => {
+            expect(wrapper.html()).toContain(mockEnglishContentDto.summary);
+        });
+
+        // Simulate language change to French
+        mockAppLanguageAsRef.value = mockLanguageDtoFra;
```
[the remark this answers is not in the document]
Comment (laurensgroeneveld) [reply app/src/pages/SingleContent.spec.ts:167]: Ivan says, check how the language switcher changes it.

[the remark this answers is not in the document]
Comment (laurensgroeneveld) [reply app/src/pages/SingleContent.spec.ts:167]: In `LanguageModel` it sets `appLanguageIdAsRef` which is not readonly, as opposed to `appLanguageAsRef`.

File: app/src/pages/SingleContent.vue [hunk 7]
```diff
@@ -114,7 +144,7 @@ onMounted(async () => {
 
         <div
             class="mt-6 border-t border-zinc-200 pt-6 dark:border-zinc-500"
-            v-if="content.tags.length > 0"
+            v-if="content.tags.length != 0"
```
Comment (laurensgroeneveld) [inline app/src/pages/SingleContent.vue:147]: Why did you change this from `>` to `!=`? It seems clearer to me if it's `>` - I would never expect it to be less than zero.

File: app/src/pages/SingleContent.vue [hunk 8]
```diff
@@ -34,6 +38,32 @@ watch(
...
+        if (appLanguageAsRef.value?._id != content.value.language) {
+            const contentDocs = await db.whereParent(content.value.parentId);
+            const preferred = contentDocs.find((c) => c.language == appLanguageAsRef.value?._id);
+
+            if (preferred) {
+                content.value = preferred;
+                await router.replace({ name: "content", params: { slug: preferred.slug } });
+                return;
+            }
+            useNotificationStore().addNotification({
```
Comment (laurensgroeneveld) [inline app/src/pages/SingleContent.vue:57]: I would still like a test for this notification. Did you look at that? Just test switching to a language for which there is no translation and then test if `addNotification` was not called:

```ts
 expect(notificationStore.addNotification).not.toHaveBeenCalled()
```

File: app/src/pages/SingleContent.vue (hunk 7 above)
Comment (belljohan3) [reply app/src/pages/SingleContent.vue:147]: @laurensgroeneveld I change like this because, in the test is not showing the tags when there is/are.

Comment (laurensgroeneveld) [reply app/src/pages/SingleContent.vue]: @belljohan3 I think then we should investigate why that happens. Logically thinking:

- There are no tags: the length is 0
- There are tags: the length is greater than 0

So checking for if it's greater than 0 should always be sufficient, because the length can't be less than 0.

The only reason why I can imagine it's causing problems in the test is that `content.tags` itself is not an array but `undefined`. That means the `.length` will cause an error to occur. But then we should rather fix the test and not change the production code. 

Making the tests work is not a goal of itself: the tests are there to check that the code is working, so the code should generally not be changed to make the tests pass.

File: app/src/pages/SingleContent.vue (hunk 8 above)
Comment (laurensgroeneveld) [reply app/src/pages/SingleContent.vue:57]: @belljohan3 I see I didn't mention it in this comment but a test that the notification _was_ called is also good I think.

Comment (laurensgroeneveld) [reply app/src/pages/SingleContent.vue:57]: This is still missing, we only test that when the language does change the notification also is displayed, please also add a test that when the language doesn't change because it isn't available the notification is displayed.

**Review by laurensgroeneveld — APPROVED** [review]
